2010-08-02  Todd Veldhuizen <todd.veldhuizen@logicblox.com>

	PR ld/11843
	* elflink.c (compute_bucket_count):  Limit try count to 100.

diff --git a/bfd/elflink.c b/bfd/elflink.c
index e0fbc83..ba84ef8 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -5374,6 +5374,7 @@ compute_bucket_count (struct bfd_link_info *info ATTRIBUTE_UNUSED,
       const struct elf_backend_data *bed = get_elf_backend_data (dynobj);
       unsigned long int *counts;
       bfd_size_type amt;
+      bfd_size_type no_improvement_count = 0;
 
       /* Possible optimization parameters: if we have NSYMS symbols we say
 	 that the hashing table must at least have NSYMS/4 and at most
@@ -5458,7 +5459,10 @@ compute_bucket_count (struct bfd_link_info *info ATTRIBUTE_UNUSED,
 	    {
 	      best_chlen = max;
 	      best_size = i;
+	      no_improvement_count = 0;
 	    }
+	  else if (++no_improvement_count == 100)
+	    break;
 	}
 
       free (counts);
